books/forms.py

```diff
@@ -0,0 +1,183 @@
+from django.utils.translation import gettext_lazy as _
+from captcha.fields import ReCaptchaField
+from captcha.widgets import ReCaptchaV2Checkbox
+from django import forms
+from .models import Book
+
+
+class BookForm(forms.ModelForm):
+    # captcha = ReCaptchaField(widget=ReCaptchaV2Checkbox)
+
+    def clean_title(self):
+        if self.cleaned_data["title"] == "":
+            raise forms.ValidationError(
+                _("Title is Required. Please ensure Book was selected from dropdown")
+            )
+
+        return self.cleaned_data["title"]
+
+    def clean_author(self):
+        if self.cleaned_data["author"] == "":
+            raise forms.ValidationError(
+                _("Author is Required. Please ensure Book was selected from dropdown")
+            )
+        return self.cleaned_data["author"]
+
+    def clean_isbn(self):
+        if self.cleaned_data["isbn"] == "":
+            raise forms.ValidationError(
+                _("ISBN is Required. Please ensure Book was selected from dropdown")
+            )
+        return self.cleaned_data["isbn"]
+
+    def clean_source(self):
+        source = self.cleaned_data["source"]
+
+        if not (source == Book.Sources.CHAT or source == Book.Sources.ATRIOC):
+            raise forms.ValidationError(_("Recommendation MUST be from CHAT or ATRIOC"))
+
+        return source
+
+    def clean_submitter(self):
+        return self.cleaned_data["submitter"]
+
+    def clean_stream_link(self):
+        cleaned_data = self.cleaned_data
+
+        stream_link = cleaned_data["stream_link"]
+        if stream_link != "" and not any(
+            url in stream_link
+            for url in [
+                "https://clips.twitch.tv/",
+                "https://www.twitch.tv/atrioc/clip/",
+            ]
+        ):
+            raise forms.ValidationError(
+                _(
+                    "Invalid stream link. The link must contain 'https://www.twitch.tv/atrioc/clip/' or 'https://clips.twitch.tv/'"
+                )
+            )
+
+        return stream_link
+
+    def clean_atrioc_streamlink(self):
+        cleaned_data = self.cleaned_data
+
+        if (
+            "source" in cleaned_data
+            and cleaned_data["source"] == Book.Sources.ATRIOC
+            and "stream_link" in cleaned_data
+            and cleaned_data["stream_link"] == ""
+        ):
```

Comment:
these conditions are a little nasty, could we shorten this, and in other places too?

Reply:
Will Address in Future PR